### PR TITLE
Adrian salazar/accounts functionality

### PR DESCRIPTION
**Pull request by AqSlz**
I integrated the existing accounts page with the backend. Now admins with "Professor" accounts can see accounts and manage them. 

# Description

The relevant files are currently under the `src/pages/accounts` folder.

In `src/pages/accounts` you we can see that the main changes are from:

- `/AdminAccountsPage.sagas.ts`. All the requests for add/edit/delete of accounts are here. Small changes pending.
- `/AdminAccountsPage.slice.ts`. State of accounts, add/edit/delete of account after confirmation from backend.

In `src/pages/accounts` you we can see some changes to these files

- `/AccountsTable.tsx`, Table now displays accounts from database.
- `/NewAccountDialog.tsx`, Now validates form content and sends new account to database
- `/AccountDialog.tsx`, Now handles changes to account information. Sends edited account to database
- `/AccountInfo.tsx`, Now can copy contents to clipboard, notifies user when that happens
- `/AccountInfoEdit.tsx` Now validates form of edited information. Displays helper text for errors.
- `/AdminActions.tsx` Actions are now functional. Additional checks for actions.
- `types.ts ` a file for the interfaces used in `src/pages/accounts`, some typescript interfaces were added.

# Motivation and Context

Adding the ability for the professor to manage accounts on the frontend

# How Has This Been Tested?

Using jest, I have mocked backend responses and tested functions on `/AdminAccountsPage.tsx`.

# Screenshots (if appropriate):

I do not want to put a .gif since there are live emails among the accounts... 

## Account Management Page

**Accounts with no name should not be there, and they won't be there after we remove them**
image

## Account Summary Dialog
**Default View**
image

**Changing Roles - TA**
image

**Deleting accounts**
image

**Editing Accounts**
image

**Creating Accounts**
image

### CODE REVIEW

File: src/pages/accounts/components/NewAccountDialog.tsx [hunk 1]
```diff
@@ -1,19 +1,39 @@
 import React from "react";
 import { Grid, Chip, Typography, TextField, MenuItem } from "@mui/material";
+import { requestNewAccount } from "../AdminAccountsPage.slice";
+import { useAppDispatch } from "../../../app/common/hooks";
 import { INewAccount, rolesEnum } from "../types";
 import Dialog from "../../../shared/GenericDialog";
 
+function isBlank(str: string) {
```
Comment (joshuadargan) [inline src/pages/accounts/components/NewAccountDialog.tsx:8]: I think we have multiple of these functions. Should we put this in a shared file?

File: src/pages/accounts/components/accountDialog/AccountInfoEdit.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
+import React from "react";
+import { Grid, Typography, TextField, MenuItem } from "@mui/material";
+import { useAppSelector } from "../../../../app/common/hooks";
+import { rolesEnum } from "../../types";
+
+function isBlank(str: string) {
```
Comment (joshuadargan) [inline src/pages/accounts/components/accountDialog/AccountInfoEdit.tsx:6]: Same as comment above

**Review by tanner-ropp — DISMISSED** [review]
LGTM

File: src/pages/accounts/AdminAccountsPage.sagas.ts [hunk 3]
```diff
@@ -1,28 +1,128 @@
...
+    yield put(setCurrentAccount(currentAccount as IAccount));
+  } catch (e: any) {
+    const data: IRequestMessage = e.response.data;
+    const feedback: IRequestMessage = data.message ? data : e;
+    yield put(requestAccountsFail(feedback));
+  }
+}
+
+function* handleNewAccountRequest(action: PayloadAction<INewAccount>): any {
+  let accountsRequest = () => adminAPI.post("/v1/user/create", action.payload);
```
Comment (joshuadargan) [inline src/pages/accounts/AdminAccountsPage.sagas.ts:51]: nit: Wouldn't hurt to add "async () => ..."

You can test if it is currently running asynchronously by throwing a print statement in the function and a print statement after it is called I think. I'm just interested to see how it works.

File: src/pages/accounts/AdminAccountsPage.sagas.ts [hunk 4]
```diff
@@ -1,28 +1,128 @@
...
+  IAccountsGET,
+  IAccountPOST,
+  IAccountDELETE,
+} from "./types";
+import { IRequestMessage } from "../../shared/types";
 
 function* handleGetAccountsRequest(): any {
-  // request to be functional next PR
-  let accountsRequest = () => adminAPI.get<IAccount[]>("/v1/user/accounts");
+  let accountsRequest = () => adminAPI.get("/v1/user/getUsers");
```
Comment (joshuadargan) [inline src/pages/accounts/AdminAccountsPage.sagas.ts:32]: nit: Wouldn't hurt to add "async () => ..."

You can test if it is currently running asynchronously by throwing a print statement in the function and a print statement after it is called I think. I'm just interested to see how it works.

File: src/pages/accounts/AdminAccountsPage.sagas.ts [hunk 5]
```diff
@@ -1,28 +1,128 @@
...
+
+function* handleModifyAccountRequest(action: PayloadAction<IAccount>): any {
+  const body: IAccount = {
+    name: action.payload.name,
+    role: action.payload.role,
+    username: action.payload.username,
+    _id: action.payload._id,
+  };
+
+  let updateRequest = () => adminAPI.put("/v1/user/updateUser", body);
```
Comment (joshuadargan) [inline src/pages/accounts/AdminAccountsPage.sagas.ts:80]: nit: Wouldn't hurt to add "async () => ..."

You can test if it is currently running asynchronously by throwing a print statement in the function and a print statement after it is called I think. I'm just interested to see how it works.

File: src/pages/accounts/AdminAccountsPage.sagas.ts [hunk 6]
```diff
@@ -1,28 +1,128 @@
...
+
+function* handleDeleteAccountRequest(action: PayloadAction<IAccount>): any {
+  // backend wants username in the body
+  // below is how to set the body in axios.delete
+  const config: AxiosRequestConfig = {
+    data: {
+      username: action.payload.username,
+    },
+  };
+  let deleteRequest = () => adminAPI.delete("/v1/user/deleteUser", config);
```
Comment (joshuadargan) [inline src/pages/accounts/AdminAccountsPage.sagas.ts:102]: nit: Wouldn't hurt to add "async () => ..."

You can test if it is currently running asynchronously by throwing a print statement in the function and a print statement after it is called I think. I'm just interested to see how it works.

**Review by joshuadargan — COMMENTED** [review]
some small nits

File: src/pages/accounts/AdminAccountsPage.sagas.ts (hunk 6 above)
Comment (AqSlz) [reply src/pages/accounts/AdminAccountsPage.sagas.ts:102]: Can be tested outside of this PR at any time

File: src/pages/accounts/components/NewAccountDialog.tsx (hunk 1 above)
Comment (AqSlz) [reply src/pages/accounts/components/NewAccountDialog.tsx]: Imported from where it existed before

Comment (AqSlz) [reply src/pages/accounts/components/NewAccountDialog.tsx]: Added file for shared functions

**Review by tanner-ropp — APPROVED** [review]

**Pull request merged**